clash-lib/src/Clash/Core/Evaluator.hs

```diff
@@ -640,53 +479,39 @@ substAlt dc tvs xs args e = substTm "Evaluator.substAlt" subst e
   subst0     = mkSubst (mkInScopeSet inScope)
 
 -- | Allocate let-bindings on the heap
-allocate :: Heap -> Stack -> [LetBinding] -> Term -> State
-allocate (Heap gh gbl h ids is0) k xes e =
-  (Heap gh gbl (h `extendVarEnvList` xes') ids' isN,k,e')
+allocate :: [LetBinding] -> Term -> Machine -> Machine
+allocate xes e m =
+  m { mHeapLocal = extendVarEnvList (mHeapLocal m) xes'
+    , mSupply = ids'
+    , mScopeNames = isN
+    , mTerm = e'
+    }
  where
-  xNms     = map fst xes
-  is1      = extendInScopeSetList is0 xNms
-  (ids',s) = mapAccumL (letSubst h) ids xNms
-  (nms,s') = unzip s
-  isN      = extendInScopeSetList is1 nms
-  subst    = extendIdSubstList subst0 s'
-  subst0   = mkSubst (foldl' extendInScopeSet is1 nms)
-  xes'     = zip nms (map (substTm "Evaluator.allocate0" subst . snd) xes)
-  e'       = substTm "Evaluator.allocate1" subst e
+  xNms      = fmap fst xes
+  is1       = extendInScopeSetList (mScopeNames m) xNms
+  (ids', s) = mapAccumL (letSubst (mHeapLocal m)) (mSupply m) xNms
+  (nms, s') = unzip s
+  isN       = extendInScopeSetList is1 nms
+  subst     = extendIdSubstList subst0 s'
+  subst0    = mkSubst (foldl' extendInScopeSet is1 nms)
+  xes'      = zip nms (fmap (substTm "Evaluator.allocate0" subst . snd) xes)
+  e'        = substTm "Evaluator.allocate1" subst e
 
 -- | Create a unique name and substitution for a let-binder
 letSubst
   :: PureHeap
   -> Supply
   -> Id
-  -> ( Supply
-     , (Id,(Id,Term)))
+  -> (Supply, (Id, (Id, Term)))
 letSubst h acc id0 =
   let (acc',id1) = uniqueInHeap h acc id0
   in  (acc',(id1,(id0,Var id1)))
-
--- | Create a name that's unique in the heap
-uniqueInHeap
-  :: PureHeap
-  -> Supply
-  -> Id
-  -> (Supply, Id)
-uniqueInHeap h ids x = case lookupVarEnv x' h of
-  Just _ -> uniqueInHeap h ids' x
-  _ -> (ids',x')
- where
-  (i,ids') = freshId ids
-  x'       = modifyVarName (\nm -> nm {nameUniq = i}) x
-
-wrapUnsigned :: Integer -> Integer -> Integer
-wrapUnsigned n i = i `mod` sz
  where
-  sz = 1 `shiftL` fromInteger n
+  -- | Create a name that's unique in the heap
+  uniqueInHeap :: PureHeap -> Supply -> Id -> (Supply, Id)
```

Comment:
`mkUniqueHeapId`?

Reply:
I didn't actually define this function, I just moved it around. I'm actually not happy with the whole `letSubst` function (the type feels a little wacky for what it's doing now there is more abstraction)